packages/libs/avector/src/lib.rs

```diff
@@ -0,0 +1,551 @@
+//! Concurrent Read Optimized Vector
+//!
+//! Vector which allows concurrent read access, simultaneous to writing.
+//! The vector can only be manipulated through pushing of items, once an item is pushed it can no
+//! longer be accessed mutably.
+//!
+//! While reading is concurrent and can happen while values are being written,
+//! values can only be written synchronously and are guarded by a spinlock.
+//! This means that [`AVec`] works best when contention is low and values are mostly being
+//! read.
+//!
+//! ## How it works
+//!
+//! [`AVec`] is based on two parts: length (through an atomic) and an intrusive list of
+//! arrays which store items.
+//! The size of a bucket can be controlled through the const generic `N`, allowing one to tradeoff
+//! speed vs. allocated space.
+//!
+//! The inner workings are best explained through a demonstration:
+//!
+//! ```text
+//! Step 1:
+//!     Vec<u8, 4>: [_, _, _, _]
+//!     A: Acquire Lock
+//!     B: Read, Length: 0
+//!
+//! Step 2:
+//!     Vec<u8, 4>: [_, _, _, _]
+//!     A: Load Length (0)
+//!     B: .next(): None
+//!
+//! Step 3:
+//!     Vec<u8, 4>: [1, _, _, _]
+//!     A: Push Value
+//!     B: Read, Length: 0
+//!
+//! Step 4:
+//!     Vec<u8, 4>: [1, _, _, _]
+//!     A: Store Length + 1
+//!     B: .next(): None
+//!
+//! Step 5:
+//!     Vec<u8, 4>: [1, _, _, _]
+//!     B: .iter(), Length: 1
+//! ```
+//!
+//! As one can see, the length is only changed, once the push is done, this means that even though
+//! we mutate and read the array at the same time we can guarantee, that never an uninitialized
+//! **or** write on the same address, entry will take place.
+//! Only once we can guarantee that writing to an entry has been done, and will never take place we
+//! will increment the length.
+//!
+//! Once a bucket has filled up, the vec will allocate a new bucket on the heap and use that,
+//! creating a linked list (into one direction).
+//! This means that every `N`th insertion will be slightly slower, as an allocation needs to be
+//! performed.
+#![no_std]
+#![warn(
+    unreachable_pub,
+    clippy::pedantic,
+    clippy::nursery,
+    clippy::undocumented_unsafe_blocks
+)]
+// TODO: once more stable introduce: warning missing_docs
+#![allow(clippy::module_name_repetitions)]
+#![allow(clippy::redundant_pub_crate)]
+
+extern crate alloc;
+
+use alloc::boxed::Box;
+use core::{cell::UnsafeCell, mem::MaybeUninit, ops::Index, ptr};
+
+use crate::{
+    lock::AtomicLock,
+    sync::{AtomicUsize, Ordering},
+};
+
+mod lock;
+pub(crate) mod sync;
+
+// note(bm): [PERFORMANCE] increase from 16 to 64
```

Comment:
We chose to not include names in comments, git diffs are powerful enough and a name on a line can get outdated easily.
```suggestion
// note: [PERFORMANCE] increase from 16 to 64
```